source/extensions/config_subscription/grpc/grpc_mux_failover.h

```diff
@@ -30,19 +30,23 @@ namespace Config {
  * will be followed by the GrpcMuxFailover calling the GrpcStreamCallbacks on the GrpcMux
  * object that initialized it.
  *
- * Note: this class is WIP and should be considered alpha!
- * At the moment, the supported policy is as follows:
+ * To simplify the state-machine, Envoy can be in one of the mutually exclusive states:
+ *   ConnectingToPrimary - attempting to connect to the primary source.
+ *   ConnectedToPrimary - after receiving a response from the primary source.
+ *   ConnectingToFailover - attempting to connect to the failover source.
+ *   ConnectedToFailover - after receiving a response from the failover source.
+ *   None - not attempting to connect or connected to any source (e.g., upon  initialization).
+ *
  * The GrpcMuxFailover attempts to establish a connection to the primary source. Once a response is
  * received from the primary source it will be considered available, and the failover will not be
  * used. Any future reconnection attempts will be to the primary source only.
  * However, if no response is received from the primary source, and accessing the primary has
  * failed 2 times in a row, the GrpcMuxFailover will attempt to establish a connection to the
- * failover source. If a response from the failover source is received, only the failover source
- * will be used.
- * If the failover source is unavailable, the GrpcMuxFailover will alternate between attempts
- * to reconnect to the primary source and the failover source.
- * In the future, this behavior may change, and the GrpcMuxFailover will always
- * prefer the primary source, even if prior connection to the failover was successful.
+ * failover source. If a response from the failover source is received, it will be the source of
+ * configuration until the connection is closed. In case the connection is closed, Envoy will
+ * attempt to connect to the primary, before retrying to connect to the failover source.
+ * If the failover source is unavailable or a connection to it is closed, the GrpcMuxFailover
+ * will alternate between attempts to reconnect to the primary source and the failover source.
```

Comment:
Is this true on startup a well?

Reply:
Yeah... I tried to rephrase this comment. Feel free to point out if it should be further improved.